keps/sig-autoscaling/draft-20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,170 @@
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+This proposal aims at allowing Pod resource requests & limits to be updated
+in-place, without a need to restart the Pod or its Containers.
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
+Resources, denoting **desired** resources.
+Additionally PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not enough to handle it,
+* load has decreased significantly and currently allocated resources are unused
+  and thus wasted,
+* Resources have simply been set improperly.
+
+Currently changing Resources allocation requires the Pod to be recreated since
+the PodSpec is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating a Pod nor restarting a
+Container addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource update is not available.
+* Secondary: allow users to specify which Pods and Containers can be updated
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed an in-place resource update. These cases should be handled by
+actors which initiated the update.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to resources and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodStatus is extended with information about actually allocated resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration,
+  denoting **desired** state of the Pod,
```

Comment:
For bind, I believe the intent is for scheduler continue to use PodSpec.Container.ResourceRequirements to find a node that meets PodFitsResources predicate check, but set PodStatus.ContainerStatus.ResourceAllocated to the values in ResourceRequirements at binding, and use ResourceAllocated values for cache accounting instead of ResourceRequirements. The kubelet looks at PodStatus.ContainerStatus.ResourceAllocated to set the limits.  

For Update, scheduler looks at the desired values in PodSpec.Container.ResourceRequirements to see if In-place resize is possible, and Update PodStatus.ContainerStatus.ResourceAllocated if yes, reschedule if not and allowed by policy, or fail and let controller retry / VPA handle it.

@kgolab please correct me if I'm mistaken.

Reply:
@vinaykul Yes, I remember that algorithm from a previous discussion with @kgolab and it is fine. I just wanted to ensure that we don't confuse users. We can clarify that the scheduler still considers resource requirements as a constraint at scheduling time.

@kgolab Yes, clarifying that it is still a constraint for scheduling helps.

Reply:
The updated design does not auto-reschedule, and it is up to the initiating actor to evict if in-place resize is not possible. So resolving this conversation, please unresolve if anything is still outstanding.